kernel/frame_allocator/src/lib.rs

```diff
@@ -262,113 +278,287 @@ pub enum MemoryRegionType {
+/// the `page_table_entry` crate, since `page_table_entry` must depend on types
+/// from this crate in order to enforce safety when modifying page table entries.
+pub(crate) fn into_allocated_frames(frames: FrameRange) -> Frames<{FrameState::Unmapped}> {
+    let typ = if contains_any(&RESERVED_REGIONS.lock(), &frames) {
+        MemoryRegionType::Reserved
+    } else {
+        MemoryRegionType::Free
+    };
+    Frames::new(typ, frames)
+}
+
+impl<const S: FrameState> Drop for Frames<S> {
+    fn drop(&mut self) {
+        match S {
+            FrameState::Unmapped => {
+                if self.size_in_frames() == 0 { return; }
+        
+                let unmapped_frames: Frames<{FrameState::Unmapped}> = Frames {
+                    typ: self.typ,
+                    frames: self.frames.clone(),
+                };
+        
+                let list = if unmapped_frames.typ == MemoryRegionType::Reserved {
+                    &FREE_RESERVED_FRAMES_LIST
+                } else {
+                    &FREE_GENERAL_FRAMES_LIST
+                };
+        
+                // Simply add the newly-deallocated chunk to the free frames list.
+                let mut locked_list = list.lock();
+                let res = locked_list.insert(unmapped_frames);
+                match res {
+                    Ok(_inserted_free_chunk) => (),
+                    Err(c) => error!("BUG: couldn't insert deallocated chunk {:?} into free frame list", c),
+                }
+                
+                // Here, we could optionally use above `_inserted_free_chunk` to merge the adjacent (contiguous) chunks
+                // before or after the newly-inserted free chunk. 
+                // However, there's no *need* to do so until we actually run out of address space or until 
+                // a requested address is in a chunk that needs to be merged.
+                // Thus, for performance, we save that for those future situations.
+            }
+            FrameState::Mapped => panic!("We should never drop a mapped frame! It should be forgotten instead."),
+        }
     }
 }
 
-impl AllocatedFrames {
-    /// Returns an empty AllocatedFrames object that performs no frame allocation. 
-    /// Can be used as a placeholder, but will not permit any real usage. 
-    pub const fn empty() -> AllocatedFrames {
-        AllocatedFrames {
-            frames: FrameRange::empty()
+impl<'f> IntoIterator for &'f Frames<{FrameState::Unmapped}> {
+    type IntoIter = UnmappedFramesIter<'f>;
+    type Item = UnmappedFrame<'f>;
+    fn into_iter(self) -> Self::IntoIter {
+        UnmappedFramesIter {
+            _owner: self,
+            range: self.frames.iter(),
+        }
+    }
+}
+
+/// An iterator over each [`UnmappedFrame`] in a range of [`Frames<{FrameState::Unmapped}>`].
+/// 
+/// To Do (get Kevin's input): Description is no longer valid, since we have an iterator for RangeInclusive now.
```

Comment:
I think this comment can be made smaller, just say that we implement out iterator to tie the lifetime of the `UnmappedFrame` to the original object. Remove the portions about `RangeInclusive` iterator since we have our own implementation now.
Or should we keep this in, and mention that we created our own `RangeInclusive` to deal with this issue?

Reply:
It seems like you overwrote my recent changes. See the latest version from the main branch, it's more accurate.